File: docs/zdl-feed2html.php
<?php
# ZigzagDownLoader (ZDL)
# 
# This program is free software: you can redistribute it and/or modify it 
# under the terms of the GNU General Public License as published 
# by the Free Software Foundation; either version 3 of the License, 
# or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, 
# but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
# or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License 
# along with this program. If not, see http://www.gnu.org/licenses/. 
# 
# Copyright (C) 2011: Gianluca Zoni (zoninoz) <zoninoz@inventati.org>
# 
# For information or to collaborate on the project:
# https://savannah.nongnu.org/projects/zdl
# 
# Gianluca Zoni (author)
# http://inventati.org/zoninoz
# zoninoz@inventati.org
#

function getLocale () {
    $prefLocales = array_reduce(
        explode(',', $_SERVER['HTTP_ACCEPT_LANGUAGE']), 
        function ($res, $el) { 
            list($l, $q) = array_merge(explode(';q=', $el), [1]); 
            $res[$l] = (float) $q; 
            return $res; 
        }, 
        []);
    arsort($prefLocales);
    foreach($prefLocales as $key => $value) {
        if ($value == 1) {
            $lang = $key;
            break;
        }	
    }
    
    // Set default language if a `$lang` version of site is not available
    if ($lang !== "it" && $lang !== "en")
        $lang = 'en';
    return $lang;
}

function getLocaleReferer () {
    if (array_key_exists('HTTP_ORIGIN', $_SERVER)) {
        $origin = $_SERVER['HTTP_ORIGIN'];
    }
    else if (array_key_exists('HTTP_REFERER', $_SERVER)) {
        $origin = $_SERVER['HTTP_REFERER'];
    } else {
        $origin = $_SERVER['REMOTE_ADDR'];
    }
    $origin = $_SERVER['HTTP_REFERER'];

    return print_r($origin[0]);
    /* preg_match("/\/(it|en)\//", $origin, matches);   $lang = $matches[1]; */
    
    // Set default language if a `$lang` version of site is not available
    /* if ($lang !== "it" && $lang !== "en")        $lang = 'en';    return $lang; */    
}

function getLocaleParam () {
    return $_SERVER['HTTP_GET'];
}

function displayFeed($url){
    $getfile = html_entity_decode(file_get_contents($url));
    $lang = getLocaleReferer();

    $xml = new SimpleXMLElement($getfile);
    //$xml = simplexml_load_file($url);

    $feed_info = array();
    $feed_art = array();

    $feed_info['titolo_feed'] = $xml->title; 
    $feed_info['id_feed'] = $xml->id; 
    $feed_info['updated_feed'] = $xml->updated;

    // $feed_info['id_feed'] è l'url del feed, sostituito dalla pagina html di savannah
    echo "<a href='" . "https://savannah.nongnu.org/news/?group_id=11047" . "' target='_blank'>" . $feed_info['titolo_feed'] . "</a>";

    $i = 0; 
    foreach($xml->entry as $item)
    {
        if (preg_match('/^\[' . $lang . '\]/', $item->title)) {
            $feed_art[$i]['titolo_articolo'] = preg_replace('/^\[' . $lang. '\]\ /','', $item->title);
            $feed_art[$i]['descr_articolo'] = $item->content->asXML();
            $feed_art[$i]['autore_articolo'] = $item->author->name;
            $feed_art[$i]['data_articolo'] = $item->updated;
            $feed_art[$i]['link_articolo'] = $item->id;
            $i++;
        }
        elseif (!preg_match('/^\[[a-z]{2}\]/', $item->title)) {
            $feed_art[$i]['titolo_articolo'] = $item->title;
            $feed_art[$i]['descr_articolo'] = $item->content->asXML();
            $feed_art[$i]['autore_articolo'] = $item->author->name;
            $feed_art[$i]['data_articolo'] = $item->updated;
            $feed_art[$i]['link_articolo'] = $item->id;
            $i++;
        }            
    }

    $total = array_merge($feed_info,$feed_art);
    //print_r($total);

    foreach ($feed_art as $key => $value) {
        echo "<hr />";
        echo "<h3><a href='" . $value['link_articolo'] . "' target='_blank'>" . $value['titolo_articolo']."</a></h3>";
        echo "<p>" . $value['descr_articolo'] . "</p>";
        //echo "<p>by " . $value['autore_articolo'] . "</p>";
        echo "<div class='feed_item_date'>Data: ".$value['data_articolo']."</div>";
    }
}

function displayHead() {
    $header = "<html lang=\"it\">
<head>
<title>ZigzagDownLoader (ZDL)</title>
<meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-8\">
<meta name=\"description\" content=\"ZigzagDownLoader (ZDL)\">
<meta name=\"generator\" content=\"makeinfo 4.13\">
<meta http-equiv=\"Content-Security-Policy\" content=\"upgrade-insecure-requests\">
<link title=\"Top\" rel=\"start\" href=\"index.html#Top\">
<link rel=\"next\" href=\"Il-comando-ZDL.html#Il-comando-ZDL\" title=\"Il comando ZDL\">
<link href=\"https://www.gnu.org/software/texinfo/\" rel=\"generator-home\" title=\"Texinfo Homepage\">
<!--
ZigzagDownLoader (ZDL)

 This program is free software: you can redistribute it and/or modify it
 under the terms of the GNU General Public License as published
 by the Free Software Foundation; either version 3 of the License,
 or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program. If not, see http://www.gnu.org/licenses/.

 Copyright (C) 2011
 Gianluca Zoni <<zoninoz@inventati.org>>

 For information or to collaborate on the project:
 `https://savannah.nongnu.org/projects/zdl'

 Gianluca Zoni (author)
 `https://inventati.org/zoninoz'
 <zoninoz@inventati.org>-->
<link rel=\"stylesheet\" type=\"text/css\" href=\"https://www.nongnu.org/zdl/zdl_rss_style.css\">
</head>
<body>";
          echo $header;
}

function displayTail() {
    echo "</body></html>";
}

//displayHead();
//displayFeed("https://savannah.nongnu.org/news/atom.php?group=zdl");
echo getLocaleParam();
//displayTail();

?>

File: docs/zdl-feed2html.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

print("Access-Control-Allow-Origin: *")
print("Content-Security-Policy: frame-ancestors *")
print("Content-Type: text/html; charset=utf-8\n")

import cgitb
cgitb.enable()

import urllib.request
import re
import sys
import ssl
import html
import os
from urllib.parse import parse_qs

if sys.version_info >= (3, 7):
    sys.stdout.reconfigure(encoding='utf-8')

# --- LOGICA DI SELEZIONE DELLA LINGUA (getLocaleParam) ---
# Legge il parametro ?origin= dall'URL (es. ?origin=http://nongnu.org)
query_string = os.environ.get('QUERY_STRING', '')
params = parse_qs(query_string)
origin_param = params.get('origin', [''])[0]

if "it" in origin_param:
    lang = "it"
else:
    lang = "en"

try:
    # Scarica il feed da Savannah
    FEED_URL = "https://savannah.nongnu.org/news/atom.php?group=zdl"
    req = urllib.request.Request(
        FEED_URL, 
        headers={'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0'}
    )
    
    context_ssl = ssl._create_unverified_context()
    
    with urllib.request.urlopen(req, context=context_ssl, timeout=10) as response:
        raw_data = response.read()
    
    xml_text = raw_data.decode('utf-8', errors='ignore')
    
    # --- DISPLAY HEAD ---
    print(f"""<html lang="{lang}">
<head>
<title>ZigzagDownLoader (ZDL)</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="description" content="ZigzagDownLoader (ZDL)">
<meta http-equiv="Content-Security-Policy" content="upgrade-insecure-requests">
<link rel="stylesheet" type="text/css" href="https://www.nongnu.org/zdl/zdl_rss_style.css">
</head>
<body>""")

    # Link principale al gruppo Savannah
    # print("<a href='https://savannah.nongnu.org/news/?group_id=11047' target='_blank'>ZDL News</a>")

    # Isoliamo tutte le <entry>
    entries = re.findall(r'<entry\b.*?>(.*?)</entry>', xml_text, re.DOTALL)
    
    for entry in entries:
        # 1. Estrazione del Titolo grezzo
        title_match = re.search(r'<title\b.*?>(.*?)</title>', entry, re.DOTALL)
        raw_title = title_match.group(1).strip() if title_match else "Senza titolo"
        raw_title = html.unescape(raw_title)
        
        # 2. Estrazione del Link/ID dell'articolo
        id_match = re.search(r'<id\b.*?>(.*?)</id>', entry, re.DOTALL)
        link_articolo = id_match.group(1).strip() if id_match else "#"
        
        # 3. Estrazione della Data
        updated_match = re.search(r'<updated\b.*?>(.*?)</updated>', entry, re.DOTALL)
        data_articolo = updated_match.group(1).strip() if updated_match else ""
        
        # 4. Estrazione del Contenuto HTML (descr_articolo)
        content_match = re.search(r'<content\b.*?>(.*?)</content>', entry, re.DOTALL)
        descr_articolo = content_match.group(1).strip() if content_match else ""
        
        # --- FILTRO LINGUA (Stessa logica del tuo if/elseif PHP) ---
        mostra_articolo = False
        titolo_filtrato = raw_title
        
        prefix = f"[{lang}]"
        if raw_title.startswith(prefix):
            # Rimuove il prefisso [it] o [en] dal titolo
            titolo_filtrato = raw_title[len(prefix):].strip()
            mostra_articolo = True
        elif not re.match(r'^\[[a-z]{2}\]', raw_title):
            # Mostra l'articolo se non ha alcun prefisso di lingua internazionale
            mostra_articolo = True
            
        # --- PRINT ARTICOLO FORMATTATO ---
        if mostra_articolo:
            print("<hr />")
            print(f"<h3><a href='{link_articolo}' target='_blank'>{titolo_filtrato}</a></h3>")
            print(f"<p>{descr_articolo}</p>")
            print(f"<div class='feed_item_date'>Data: {data_articolo}</div>")
            
    # --- DISPLAY TAIL ---
    print("</body></html>")

except Exception as e:
    print(f"<html><body><p style='color:red;'>Errore: {e}</p></body></html>")
